src/main/cpp/RLBotInterface/BallPrediction/src/bvh.cpp

```diff
@@ -0,0 +1,345 @@
+#include "..\inc\bvh.h"
+#include "..\inc\timer.h"
+#include "..\inc\morton.h"
+#include "..\inc\bit_packing.h"
+
+#include <algorithm>
+
+std::vector < uint64_t > morton_sort(
+    const std::vector < aabb > & boxes, 
+    const aabb & global) {
+
+  constexpr size_t dim = 3;
+
+  uint32_t num_boxes = uint32_t(boxes.size());
+  float x_offset = global.min_x;
+  float y_offset = global.min_y;
+  float z_offset = global.min_z;
+
+  int b = bits_needed(num_boxes);
+  int bits_per_dimension = (64 - b) / dim;
+  int divisions_per_dimension = 1 << bits_per_dimension;
+
+  float scale = float(divisions_per_dimension - 1);
+
+  float x_scale = scale / (global.max_x - global.min_x);
+  float y_scale = scale / (global.max_y - global.min_y);
+  float z_scale = scale / (global.max_z - global.min_z);
+
+  std::vector < uint64_t > code_ids(num_boxes);
+
+  for (uint32_t i = 0; i < num_boxes; i++) {
+
+    aabb box = boxes[i];
+
+    // get the centroid of the ith bounding box
+    float cx = 0.5f * (box.min_x + box.max_x);
+    float cy = 0.5f * (box.min_y + box.max_y);
+    float cz = 0.5f * (box.min_z + box.max_z);
+
+    uint64_t ux = (uint64_t)((cx - x_offset) * x_scale);
+    uint64_t uy = (uint64_t)((cy - y_offset) * y_scale);
+    uint64_t uz = (uint64_t)((cz - z_offset) * z_scale);
+
+    uint64_t code = morton::encode(ux, uy, uz); 
+
+    code_ids[i] = (code << b) + uint64_t(i);
+
+  }
+
+  std::sort(code_ids.begin(), code_ids.end());
+
+  return code_ids;
+
+}
+
+// convenience class for bounds checking morton codes,
+// and computing the number of prefix bits common to two 
+// 64-bit words
+class PrefixComparator{
+
+  public:
+
+    const uint64_t   base;
+    const uint64_t * codes;
+    const int        n;
+
+    PrefixComparator(const uint64_t   _base,
```

Comment:
can you make constructor parameters consistent across the code base?
this uses _value
while the other cases just use value